src/main/java/org/opensearch/agent/tools/utils/clustering/ClusteringHelper.java

```diff
@@ -0,0 +1,525 @@
+
+            if (traceId == null || traceId.isEmpty()) {
+                throw new IllegalArgumentException("Trace ID cannot be null or empty");
+            }
+
+            if (vector == null) {
+                throw new IllegalArgumentException("Vector for trace ID '" + traceId + "' is null");
+            }
+
+            if (vector.length == 0) {
+                throw new IllegalArgumentException("Vector for trace ID '" + traceId + "' is empty");
+            }
+
+            // Ensure all vectors have the same dimension
+            if (vectorDimension == -1) {
+                vectorDimension = vector.length;
+            } else if (vector.length != vectorDimension) {
+                throw new IllegalArgumentException(
+                    "Vector dimension mismatch: expected "
+                        + vectorDimension
+                        + " but got "
+                        + vector.length
+                        + " for trace ID '"
+                        + traceId
+                        + "'"
+                );
+            }
+
+            // Check for NaN or Infinity values
+            for (int i = 0; i < vector.length; i++) {
+                if (Double.isNaN(vector[i]) || Double.isInfinite(vector[i])) {
+                    throw new IllegalArgumentException(
+                        "Vector for trace ID '" + traceId + "' contains invalid value at index " + i + ": " + vector[i]
+                    );
+                }
+            }
+        }
+    }
+
+    /**
+     * Extracts original vector indices for each K-means cluster
+     *
+     * @param clusters K-means clustering result
+     * @param vectors Original vector array
+     * @return List of clusters with original vector indices
+     */
+    private List<List<Integer>> extractClusterIndices(List<CentroidCluster<DoublePoint>> clusters, double[][] vectors) {
+        List<List<Integer>> result = new ArrayList<>();
+        for (CentroidCluster<DoublePoint> cluster : clusters) {
+            List<Integer> clusterIndices = new ArrayList<>();
+            for (DoublePoint point : cluster.getPoints()) {
+                // Find the original index of this point
+                for (int i = 0; i < vectors.length; i++) {
+                    if (Arrays.equals(vectors[i], point.getPoint())) {
+                        clusterIndices.add(i);
+                        break;
+                    }
+                }
+            }
+            if (!clusterIndices.isEmpty()) {
+                result.add(clusterIndices);
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Perform hierarchical clustering on a subset of vectors
+     *
+     * @param vectors Input vectors for clustering
+     * @param indexTraceIdMap Mapping from vector index to trace ID
+     * @return List of trace IDs representing cluster centroids
+     */
+    private List<String> performHierarchicalClustering(double[][] vectors, Map<Integer, String> indexTraceIdMap) {
```

Comment:
Seems the function is just a wrapper of `performClustering`. Can we remove it?

Reply:
sure, we can remove it.